pip/pip-274.md

```diff
@@ -0,0 +1,126 @@
+# Background knowledge
+
+Apache Pulsar is a distributed messaging system that supports multiple messaging protocols and storage methods. Among them, Pulsar Topic Compaction is a mechanism to clean up duplicate messages in topics to reduce storage space and improve system efficiency.
+More topic compaction details can be found in [Pulsar Topic Compaction](https://pulsar.apache.org/docs/en/concepts-topic-compaction/).
+
+# Motivation
+
+Currently, the implementation of Pulsar Topic Compaction is fixed and does not support custom strategy, which limits users from using more Compactor policies in their applications.
+
+
+For example, we need to parse the Kafka format then compact message in Kop, but the current implementation of Pulsar topic compaction does not support this feature.
+Another the topic compaction logic implemented in `TwoPhaseCompactor` only compacts messages to the last one, but sometimes we need to keep the first valid message e.g [`StrategicTwoPhaseCompactor`](https://github.com/coderzc/pulsar/blob/0e9935c493060b13b322a84c5418146423992369/pulsar-broker/src/main/java/org/apache/pulsar/compaction/StrategicTwoPhaseCompactor.java).
+
+So we need to make the topic compactor pluggable to support more compaction strategy.
+
+# Goals
+
+## In Scope
+
+<!--
+What this PIP intend to achieve once It's integrated into Pulsar.
+Why does it benefit Pulsar.
+-->
+
+Make the compactor pluggable.
+
+## Out of Scope
+
+<!--
+Describe what you have decided to keep out of scope, perhaps left for a different PIP/s.
+-->
+
+
+# High Level Design
+
+<!--
+Describe the design of your solution in *high level*.
+Describe the solution end to end, from a birds-eye view.
+Don't go into implementation details in this section.
+
+I should be able to finish reading from beginning of the PIP to here (including) and understand the feature and 
+how you intend to solve it, end to end.
+
+DON'T
+* Avoid code snippets, unless it's essential to explain your intent.
+-->
+
+Make the topic compactor pluggable, users can customize the compactor implementation according to their own special scenarios.
+
+
+# Detailed Design
+
+## Design & Implementation Details
+
+<!--
+This is the section where you dive into the details. It can be:
+* Concrete class names and their roles and responsibility, including methods.
+* Code snippets of existing code.
+* Interface names and its methods.
+* ...
+-->
+* Define a standard Compactor interface that specifies the methods and properties that the Compactor implementation needs to implement. This interface should include methods for Compactor initialization, Compactor execution, and getting Compactor stats.
+```java
+public interface Compactor {
+
+    void initialize(ServiceConfiguration conf,
+                    PulsarClient pulsar,
+                    BookKeeper bk,
+                    ScheduledExecutorService scheduler);
```

Comment:
It's not a good abstraction to have `BookKeeper` (or even `PulsarClient`) involved. These parameters are just copied from the existing `Compactor`'s constructor, which is not abstracted well.

Reply:
If the motivation is to handle messages whose format is other than pulsar, we might need to pass some functions like:

```java
interface CompactEntryProcessor {

    // Just an example method
    void handle(Entry entry);
}
```

and pass the processor to the compactor.

Reply:
Do you mean to abstract the `CompactEntryProcessor` and add a `compactEntryProcessorClassName` configuration?
But we still need to construct this `CompactEntryProcessor` by BookKeeper.

How about  abstraction the `CompactStorage` like:
```java
interface CompactionHandle {

    CompletableFuture<Void> asyncAddEntry(RawMessage msg);

    CompletableFuture<Void> flush();

    CompletableFuture<Void> close(Long storageId);

}

public interface CompactStorage {

    CompletableFuture<CompactionHandle> createCompactionStorage();

    void deleteCompactionStorage(CompactionHandle compactionHandle);
}

Reply:
No. You understood wrong. Adding these two interfaces is complicated for users to implement. I mean, we only need ways to handle entries of different format, so users don't need to implement the same logic of reading entries or writing entries again.

For example, we only need to replace

```java
if (RawBatchConverter.isReadableBatch(m)) {
    // ...
} else {
    // ...
}
```

with

```java
processor.handle(m, latestForKey);
```

Where the `processor` is an interface:

```java
interface CompactMessageProcessor {

    /**
     * Process the message to update the latest key => value map.
     * 
     * @param message the raw message
     * @param latestForKey the map that maps the key to the latest value
     */
    void handle(RawMessage message, Map<String, MessageId> latestForKey);
}
```

Then, we don't need to rewrite the same logic again. We only need to change the logic about how to process a `RawMessage`.